caffe2/operators/tensorrt_op_trt.cc

```diff
@@ -0,0 +1,184 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "caffe2/operators/tensorrt_op_trt.h"
+#include "caffe2/core/logging.h"
+
+#include <unordered_map>
+
+namespace caffe2 {
+
+namespace {
+// Note that input of trt tensor is in CHW format, while our tensor is NCHW
+bool CheckDims(
+    const nvinfer1::Dims& nv_dims,
+    const std::vector<TIndex>& c2_dims) {
+  std::cout << "trt_dims: ";
+  for (int i = 0; i < nv_dims.nbDims; ++i) {
+    std::cout << nv_dims.d[i] << ",";
+  }
+  std::cout << std::endl;
+  std::cout << "c2_dims: ";
+  for (int i = 0; i < c2_dims.size(); ++i) {
+    std::cout << c2_dims[i] << ",";
+  }
+  std::cout << std::endl;
+
+  if (nv_dims.nbDims + 1 != c2_dims.size()) {
+    return false;
+  }
+  for (int i = 0; i < nv_dims.nbDims; ++i) {
+    if (nv_dims.d[i] != c2_dims[i + 1]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+} // namespace
+
+// Upon construction, we build the inference enigne by deserializing from
+// protobuf string. And since we know the input/output blobs, we can do the
+// binding here too.
+TensorRTOp::TensorRTOp(const OperatorDef& operator_def, Workspace* ws)
+    : Operator<CUDAContext>(operator_def, ws),
+      logger_((nvinfer1::ILogger::Severity)(
+          OperatorBase::GetSingleArgument<int>("log_verbosity", 2))),
+      batch_size_(OperatorBase::GetSingleArgument<int>("max_batch_size", 1)) {
+  {
+    auto engine_string =
+        OperatorBase::GetSingleArgument<std::string>("serialized_engine", "");
+    CAFFE_ENFORCE(!engine_string.empty(), "Empty serialized TensorRT engine!");
+    auto trt_runtime = InferObject(nvinfer1::createInferRuntime(logger_));
+    // TODO(support trt plugin factory)
+    trt_engine_ = InferObject(trt_runtime->deserializeCudaEngine(
+        engine_string.data(), engine_string.size(), nullptr));
+  }
+
+  if(!trt_engine_) {
+    CAFFE_THROW("Cannot deserialize TensorRT engine!");
+  }
+
+  // match and bind the input/output
+  std::unordered_map<std::string, int> inputs;
+  std::unordered_map<std::string, int> outputs;
+  for (int i = 0; i < operator_def.input_size(); ++i) {
+    inputs.emplace(operator_def.input(i), i);
+    std::cout << "Adding Input: " << operator_def.input(i) << std::endl;
+  }
+  for (int i = 0; i < operator_def.output_size(); ++i) {
+    outputs.emplace(operator_def.output(i), i);
+    std::cout << "Adding Output: " << operator_def.output(i) << std::endl;
+  }
+
+
+  int num_bindings = trt_engine_->getNbBindings();
+  for (int b = 0; b < num_bindings; ++b) {
+    const auto& name = trt_engine_->getBindingName(b);
+    nv_dims_.push_back(trt_engine_->getBindingDimensions(b));
+    if (trt_engine_->bindingIsInput(b)) {
+      std::cout << "Checking TRT input: " << name << std::endl;
+      const auto it = inputs.find(name);
+      CAFFE_ENFORCE(it != inputs.end(), MakeString("Cannot find trt input: ", name));
```

Comment:
I wonder whether we should bind inputs by position here. In case somebody rewrites the net after the TRT op is create it'd stop working. In C2 we shouldn't rely on the names of inputs in the graph to do matching, in general

Reply:
Yes, the safest way is to do it in `RunOnDevice`. But I need more info on  

> In C2 we shouldn't rely on the names of inputs in the graph to do matching, in general